z3/src/datatype_builder.rs

```diff
@@ -1,117 +1,179 @@
+use std::mem;
 use std::{convert::TryInto, ptr::null_mut};
 use z3_sys::*;
-use {Context, DatatypeBuilder, DatatypeSort, DatatypeVariant, FuncDecl, Sort, Symbol};
+use {
+    Context, DatatypeAccessor, DatatypeBuilder, DatatypeSort, DatatypeVariant, FuncDecl, Sort,
+    Symbol,
+};
 
 impl<'ctx> DatatypeBuilder<'ctx> {
-    pub fn new(ctx: &'ctx Context) -> Self {
+    pub fn new<S: Into<Symbol>>(ctx: &'ctx Context, name: S) -> Self {
         Self {
             ctx,
-            variants: Vec::new(),
+            name: name.into(),
+            constructors: Vec::new(),
         }
     }
 
-    pub fn variant(mut self, name: &str, fields: &[(&str, &Sort)]) -> Self {
-        let recognizer_name_sym = Symbol::String(format!("is-{}", name)).as_z3_symbol(self.ctx);
-        let name_sym = Symbol::String(name.to_string()).as_z3_symbol(self.ctx);
-
-        assert!(fields
-            .iter()
-            .all(|(name, sort)| sort.ctx.z3_ctx == self.ctx.z3_ctx));
-
-        let mut field_names: Vec<Z3_symbol> = Vec::with_capacity(fields.len());
-        let mut field_sorts = Vec::with_capacity(fields.len());
-
-        for (name, sort) in fields {
-            field_names.push(Symbol::String(name.to_string()).as_z3_symbol(self.ctx));
-            field_sorts.push(sort.z3_sort);
+    pub fn variant(&mut self, name: &str, fields: &'ctx [(&str, DatatypeAccessor<'ctx>)]) {
+        let mut accessor_vec: Vec<(String, &DatatypeAccessor<'ctx>)> = Vec::new();
+        for (accessor_name, accessor) in fields {
+            accessor_vec.push((accessor_name.to_string(), accessor));
         }
 
-        // This is unused.
-        // Z3 expects sort_refs in Z3_mk_constructor to be valid, so create it here.
-        let mut sort_refs = Vec::new();
-        sort_refs.resize(fields.len(), 0);
-
-        let constructor = unsafe {
-            Z3_mk_constructor(
-                self.ctx.z3_ctx,
-                name_sym,
-                recognizer_name_sym,
-                fields.len().try_into().unwrap(),
-                field_names.as_ptr(),
-                field_sorts.as_ptr(),
-                sort_refs.as_mut_ptr(),
-            )
-        };
+        self.constructors.push((name.to_string(), accessor_vec));
+    }
 
-        self.variants.push((fields.len(), constructor));
-        self
+    pub fn finish(&self) -> DatatypeSort<'ctx> {
```

Comment:
This should still take `self`, since attempting to call `finish` multiple times is an error right? (I assume Z3 crashes at an assertion or something)

Reply:
Right, good catch.